### ISSUE TITLE
Should `SecretStr` be idempotent?

### ISSUE THREAD

**Issue by dmontagu** [samuelcolvin/pydantic#1044]
# Question

* Pydantic version 1.2

Currently, a `SecretStr` can't be validated as a `SecretStr`:
```python
from pydantic import BaseModel, SecretStr


class M(BaseModel):
    a: SecretStr

M(a=SecretStr("s1"))
"""
pydantic.error_wrappers.ValidationError: 1 validation error for M
a
  str type expected (type=type_error.str)
"""

# Would raise the same error:
m = M(a="s1")
M(**dict(m))
```

Is this desirable? It seems to me it would be better if instances of `SecretStr` would pass validation as `SecretStr`.

I'm not sure whether "fixing" this would count as a bug fix, or would rise to the level of a breaking change (and require waiting until v2). I'd personally prefer to consider it a bug, but could be convinced either way. In any case, I think it would be a relatively easy fix.

(Note that this currently prevents the use of `SecretStr` as a field of a `response_model` in FastAPI due to the reparsing, as described in https://github.com/tiangolo/fastapi/issues/744)

-------

More generally, I think a goal for v2 should be for all pydantic-provided types (or at least, as many as possible) to be idempotent in this way. Does anyone disagree?

**Comment by samuelcolvin** [comment]
> Is this desirable? It seems to me it would be better if instances of `SecretStr` would pass validation as `SecretStr`.

Agreed. And `SecretBytes` too of course.

**Comment by MrMrRobat** [comment]
Is there acceptable way we inherit it from `str`?

**Comment by samuelcolvin** [comment]
Yes, it's possible to inherit from `str`, see `AnyUrl`. But that's not required (and wouldn't work) in this case, since `str(SecretStr('fluffy123'))` is not idempotent.

**Comment by Atheuz** [comment]
@samuelcolvin do you have any advice on how to resolve this? I would like to do so but I'm not sure how.

**Comment by Atheuz** [comment]
@MrMrRobat @dmontagu @samuelcolvin I've created a tentative PR at #1330. 

I took inspiration from the `NameEmail` type defined here: https://github.com/samuelcolvin/pydantic/blob/master/pydantic/networks.py#L325-L350

**Comment by Atheuz** [comment]
This was resolved in #1330. I believe this issue can be closed now.

**Comment by samuelcolvin** [comment]
Thanks so much.
